caldav/match.go

```diff
@@ -0,0 +1,224 @@
+func Filter(query *CalendarQuery, cos []CalendarObject) ([]CalendarObject, error) {
+	if query == nil {
+		// FIXME: should we always return a copy of the provided slice?
+		return cos, nil
+	}
+
+	out := make([]CalendarObject, 0)
+	for _, co := range cos {
+		ok, err := Match(query.CompFilter, &co)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			continue
+		}
+
+		// TODO properties are not currently filtered even if requested
+		out = append(out, co)
+	}
+	return out, nil
+}
+
+// Match reports whether the provided CalendarObject matches the query.
+func Match(query CompFilter, co *CalendarObject) (matched bool, err error) {
+	if co.Data == nil || co.Data.Component == nil {
+		// TODO is this a thing? Should this match if comp.IsNotDefined?
+		return false, nil
+	}
+	comp := co.Data.Component
+
+	if query.Name != comp.Name {
+		return query.IsNotDefined, nil
+	}
+
+	for _, compFilter := range query.Comps {
+		match, err := matchCompFilter(compFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	for _, propFilter := range query.Props {
+		match, err := matchPropFilter(propFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func matchCompFilter(filter CompFilter, comp *ical.Component) (bool, error) {
+	var matches []*ical.Component
+
+	for _, child := range comp.Children {
+		match, err := matchCompFilterChild(filter, child)
+		if err != nil {
+			return false, err
+		} else if match {
+			matches = append(matches, child)
+		}
+	}
+	if len(matches) == 0 {
+		return filter.IsNotDefined, nil
+	}
+	return true, nil
+}
+
+func matchCompFilterChild(filter CompFilter, comp *ical.Component) (bool, error) {
+	if comp.Name != filter.Name {
+		return false, nil
+	}
+
+	var zeroDate time.Time
+	if filter.Start != zeroDate {
+		match, err := matchCompTimeRange(filter.Start, filter.End, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	for _, compFilter := range filter.Comps {
+		match, err := matchCompFilter(compFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	for _, propFilter := range filter.Props {
+		match, err := matchPropFilter(propFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func matchPropFilter(filter PropFilter, comp *ical.Component) (bool, error) {
+	// TODO: this only matches first field, can there be multiple like for CardDAV?
+	field := comp.Props.Get(filter.Name)
+	if field == nil {
+		return filter.IsNotDefined, nil
+	}
+
+	var zeroDate time.Time
+	if filter.Start != zeroDate {
+		match, err := matchPropTimeRange(filter.Start, filter.End, field)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+		for _, paramFilter := range filter.ParamFilter {
```

Comment:
Since this loop is common to both `if` branches, maybe move it one scope level up?

Reply:
I moved it up. But: according to [RFC 4791, section 9.7.2](https://datatracker.ietf.org/doc/html/rfc4791#section-9.7.2), the param filters are only present/evaluated if one of the if-branches is taken. I suppose it doesn't matter, because for any valid message, there is no semantic difference. I just think the code reads a little odd when compared to the RFC.